packages/vitest/src/node/watch-filter.ts

```diff
@@ -0,0 +1,160 @@
+import readline from 'node:readline'
+import c from 'picocolors'
+import { stdout } from '../utils'
+
+const MAX_RESULT_COUNT = 10
+const SELECTION_MAX_INDEX = 7
+const ESC = '\u001B['
+
+type FilterFunc = (keyword: string) => Promise<string[]> | string[]
+
+export class WatchFilter {
+  private filterRL: readline.Interface
+  private currentKeyword: string | undefined = undefined
+  private message: string
+  private results: string[] = []
+  private selectionIndex = -1
+  private onKeyPress?: (str: string, key: any) => void
+
+  constructor(message: string) {
+    this.message = message
+    this.filterRL = readline.createInterface({ input: process.stdin, escapeCodeTimeout: 50 })
+    readline.emitKeypressEvents(process.stdin, this.filterRL)
+    if (process.stdin.isTTY)
+      process.stdin.setRawMode(true)
+  }
+
+  public async filter(filterFunc: FilterFunc): Promise<string | undefined> {
+    stdout().write(`${c.cyan('?')} ${c.bold(this.message)} › `)
```

Comment:
(refactoring suggestion) I see this is only rendered once here and thus each rendering keeps track of this using `keywordOffset`, `printKeyword` etc...

How about including this prompt in the normal re-rendering? I think that might make some strcuture simpler and I think other library takes such approach usually.

Overall, the code would look like this:

<details><summary>Code</summary>

```ts
  public async filter(filterFunc: FilterFunc): Promise<string | undefined> {
    // -- instead of prompt only render
    // stdout().write(`${c.cyan('?')} ${c.bold(this.message)} › `)
    // ++ invoke whole rendering initially
    this.render();
    ...
  }

  private render() {
    // erase and print whole string  
    this.eraseAndPrint(this.renderImpl())
    stdout().write(`${ESC}1G`) // also needs to go left (this can be included in eraseAndPrint)
  }

  // pure render function to generate string output
  private renderImpl(): string {
    let output = `${c.cyan('?')} ${c.bold(this.message)} › ${this.currentKeyword || ''}\n`

    if (!this.currentKeyword) {
      output += 'Please input filter pattern'
    }
    else if (this.currentKeyword && this.results.length === 0) {
      output += `Pattern matches no results`
    }
    else {
      ...
    }

    return output;
  }
```

</details>

This would remove the need of `keywordOffset`, `printKeyword`, and `restoreCursor` unless I'm missing some feature `restoreCursor` is providing.

Reply:
@hi-ogawa 
Thank you for your suggestion!

I also fixed this in c2245bd.

Btw, I did not remove `printKeyword` and  `restoreCursor` because I want to set the cursor in the right place even though it is hidden normally.

Reply:
I was also wondering why cursor is hidden currently. I think it can be forcefully toggled by `${ESC}?25l` and `${ESC}?25h`, but I didn't think this was a big deal either, so this part should okay as is. Thanks!